Add resource engine interface

suskin opened the pull request:
## Overview

The render phase controller was becoming overly large and a bit hard to read, so it was about time to refactor it a little bit. Also, @muvaf's upcoming work on adding another engine type means that it'd be helpful to have a better contract for decoupling the resource engine logic from the controller flow.

This change aims to do two things:

* Introduce a new, cleaner contract for resource engines (this is also the first pass at the contract)
* Move resource engine logic out of the render controller so that it isn't cluttering things up for readers and maintainers

The commits in the PR should tell a relatively reasonable story about how this happened. But, to summarize, most of the changes are refactoring and moving the existing logic behind the interface. There's also a minor change to inject the name of the render controller's stack configuration in at the time it's created; previously it was being assumed to be the same as the claim that was being processed, which isn't going to work for real use-cases.

## Testing done

I've been testing this locally similar to the previous PRs. The gist of it is:

* Create a StackConfiguration by hand
* Create an object mentioned in that StackConfiguration (I've been using `HelmChartInstall`)
* `make run`
* Check that the jobs ran successfully and objects were created

Pretty soon I plan to investigate the [integration testing offered by controller-runtime](https://godoc.org/sigs.k8s.io/controller-runtime#hdr-Testing), but I haven't quite gotten around to it yet.

## Future work

As was noted in an earlier PR, there is some hardening that will be needed for the render controller. None of it is included in this PR.

* [ ] Only create one render controller per CRD. Currently one per CRD is created for every Reconcile call in the setup controller.
* [ ] Stop render controllers when they're no longer needed.
* [ ] Only create resources when it's needed, rather than blindly on every reconcile.

suskin committed:
render phase: add resource engine interface

As part of being able to use multiple different engines, we want to be
able to have a common contract across all resource engines (such as
helm2, helm3, kustomize). This change adds an initial interface for
that, and ports some of the helm2 engine functionality to be behind the
interface.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

suskin committed:
resource engines: move engines into separate package

Now that we have an interface for resource engines, we want to move them
into their own package so that the code's a bit more organized.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

suskin committed:
render controller: move hook execution logic into resource engine

Now that we have the resource engine interface defined and split out
into its own package, and a first helm2 resource engine, we want to
execute the hook logic by using the helm2 resource engine. This moves
all of that functionality into the engine and sets up the render phase
controller to use it from the engine.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

suskin committed:
render controller: take config name on creation

Previously, the render controller was fetching the stack configuration
based on the name of the claim that it was reconciling. Now, it takes in
the name of the stack configuration when it's created, because an
individual stack controller always fetches the same stack configuration,
and grabbing the name from the claim was a bit of a temporary measure to
get things up and running.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

suskin commented:
/cc @muvaf

suskin committed:
resource engines: explain what a resource engine runner does

It seemed like explaining the semantic expectations of a resource engine
runner would be helpful.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

muvaf commented:
@suskin https://github.com/suskin/stack-template-engine/pull/4 merges the initial kustomize support to this branch.

suskin committed:
helm2 engine: assume full relative resource path

Previously, the path to the resources was assumed to be within the
`resources` folder in the stack's contents. However, it seems more
natural to put the resources in the `resources` root when there's only
one set of resource files, so this changes the helm2 engine to assume
the `resources` directory is specified as part of the path.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

suskin committed:
stack image: fix image name, fix source bundling

The image for the stack was specific to a particular resource engine,
and we don't want that, so the name is now more generic. Also, the
docker image was not including all of the source files (the new
`engines/` folder was left out), so this fixes that too.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

suskin committed:
test: add helm2 integration test

We want to be able to easily run integration tests, and for other
contributors to be able to discover how to run them. This adds a very
lightweight hook for running integration test steps, and adds a first
pass at such a test for the helm2 engine.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

suskin committed:
integration testing: explain that the stack needs to be built

The testing instructions were missing a step, which is that the stack
needs to be built before the test runs.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

suskin commented:
This PR is superseded by a [similar PR in Crossplane](https://github.com/crossplaneio/crossplane/pull/1131). The Crossplane PR uses work from this branch, so I'm going to merge these changes into master.
